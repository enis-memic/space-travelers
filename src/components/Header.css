.header {
  display: flex;
  flex-direction: row;
  margin: 20px;
  justify-content: space-between;
  align-items: center;
  border-bottom: 2px solid gainsboro;
  padding-bottom: 40px;
}

.header-1 {
  display: flex;
  justify-content: flex-start;
  margin: 0;
}

.header h1 {
  font-size: 50px;
  margin-left: 20px;
}

img {
  width: 15%;
}

.header-photo {
  display: flex;
  gap: 30px;
}
