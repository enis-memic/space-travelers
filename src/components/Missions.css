ul {
  list-style: none;
}

.tbody td {
  border: 1px solid gray;
}

.mission-list:nth-child(odd) {
  background-color: #f2f2f2;
}

.mtable {
  margin-left: 2.5%;
  margin-right: 2.5%;
  text-align: left;
  border: 1px solid gray;
  border-collapse: collapse;
}

.mission-title {
  font-weight: bold;
  text-align: left;
  padding-left: 5px;
}

.mission-description {
  width: 70%;
  text-align: left;
  padding-left: 5px;
  padding-bottom: 25px;
}

.member {
  background-color: #18a2b8;
  color: #fff;
  border: none;
  border-radius: 6px;
  font-weight: bold;
  padding: 5px;
  width: 110px;
}

.non-member {
  color: #fff;
  background-color: darkgray;
  border: none;
  border-radius: 6px;
  font-weight: bold;
  padding: 5px;
  width: 120px;
}

.Reserve {
  background: none;
  border-color: #36454f;
  color: #36454f;
  border-radius: 6px;
  padding: 10px;
  font-size: 15px;
  cursor: pointer;
  width: 100px;
  height: 60px;
}

.Reserved {
  background: none;
  border-color: red;
  color: red;
  border-radius: 6px;
  padding: 10px;
  font-size: 15px;
  cursor: pointer;
  width: 100px;
  height: 60px;
}

.Reserve:hover {
  background-color: lightgreen;
  transform: scale(1.05);
}

.Reserved:hover {
  transform: scale(1.05);
}

.mission-button {
  text-align: center;
}

.mission-status {
  text-align: center;
}

.join-btn {
  background-color: transparent;
  color: #4b5056;
  padding: 0.8rem;
  border-radius: 5px;
  border: #4b5056 1px solid;
}

.active-btn {
  background-color: transparent;
  color: #de4150;
  border: #de4150 1px solid;
  border-radius: 5px;
  padding: 0.8rem;
}
