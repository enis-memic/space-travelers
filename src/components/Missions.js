import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  fetchMissions,
  joinMission,
  cancelMission,
} from '../redux/missions/missionsSlice';

const Missions = () => {
  const { missions, fetched } = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    if (!fetched) {
      dispatch(fetchMissions());
    }
  }, [dispatch, fetched]);

  const reserveMission = (missionId) => {
    dispatch(joinMission(missionId));
  };
  const unresMission = (missionId) => {
    dispatch(cancelMission(missionId));
  };
  return (
    <table>
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr className="mission-list" key={mission.mission_id}>
            <td className="mission-title">{mission.mission_name}</td>
            <td className="mission-description">{mission.description}</td>
            <td className="mission-button">
              {mission.joined ? (
                <button
                  id={mission.mission_id}
                  type="button"
                  onClick={() => unresMission(mission.mission_id)}
                >
                  Leave Mission
                </button>
              ) : (
                <button
                  id={mission.mission_id}
                  type="button"
                  onClick={() => reserveMission(mission.mission_id)}
                >
                  Join mission
                </button>
              )}
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default Missions;
