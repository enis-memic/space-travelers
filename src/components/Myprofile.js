import { useSelector } from 'react-redux';
import './myprofile.css';

const Myprofile = () => {
  const { rockets } = useSelector((state) => state.rockets);

  const { missions } = useSelector((state) => state.missions);

  return (
    <div className="my-profile-div">
      <div>
        <h2 className="my-profile-h2">My missions</h2>
        <table className="mprtable">
          <tbody>
            {missions
              .filter((mission) => mission.joined)
              .map((joinedMission) => (
                <tr key={joinedMission.mission_id}>
                  <td className="mprtd">{joinedMission.mission_name}</td>
                </tr>
              ))}
          </tbody>
        </table>
      </div>
      <div className="mprdiv">
        <h2 className="my-profile-h2">My rockets</h2>
        <table className="mprtable">
          <tbody>
            {rockets
              .filter((rocket) => rocket.reserved)
              .map((reservedRocket) => (
                <tr key={reservedRocket.id}>
                  <td className="mprtd">{reservedRocket.rocket_name}</td>
                </tr>
              ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default Myprofile;
