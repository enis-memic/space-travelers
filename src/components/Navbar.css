a {
  text-decoration: none;
  font-size: 25px;
  color: #298cff;
}

.links {
  display: flex;
  gap: 25px;
}

.active {
  text-decoration: underline;
}

.line {
  background-color: #414548;
  width: 1px;
  height: 25px;
}

.navbar {
  margin-right: 30px;
}
