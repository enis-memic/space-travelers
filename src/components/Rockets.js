import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { fetchRockets, bookRockets, unBookRockets } from '../redux/rockets/rocketsSlice';
import './rocketscss.css';

const Rockets = () => {
  const { rockets, fetched } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!fetched) {
      dispatch(fetchRockets());
    }
  }, [dispatch, fetched]);

  const handleClick = (rocketId) => {
    dispatch(bookRockets(rocketId));
  };

  const handleReject = (rocketId) => {
    dispatch(unBookRockets(rocketId));
  };

  return (
    <div className="rockets">
      <ul className="rul">
        {rockets.map((rocket) => (
          <li className="rli" key={rocket.id}>
            <img className="rimg" src={rocket.flickr_images[0]} alt={rocket.rocket_name} />
            <div className="rdiv">
              <h4 className="rh4">{rocket.rocket_name}</h4>
              <p className="rp">
                {rocket.reserved && <span className="rspan">Reserved</span>}
                {' '}
                {rocket.description}
              </p>
              {rocket.reserved ? (
                <button className="rcancel" type="button" onClick={() => handleReject(rocket.id)}>
                  Cancel Reservation
                </button>
              ) : (
                <button className="rreserve" type="button" onClick={() => handleClick(rocket.id)}>
                  Reserve Rocket
                </button>
              )}
            </div>
          </li>
        ))}
      </ul>
    </div>

  );
};

export default Rockets;
