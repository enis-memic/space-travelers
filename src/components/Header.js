import planet from '../img/planet.png';
import Navbar from './Navbar';
import './Header.css';

const Header = () => (
  <div className="header">
    <div className="header-1">
      <img src={planet} alt="planet logo" />
      <h1 className="header-photo">Space Travelers&apos; Hub</h1>
    </div>
    <Navbar />
  </div>
);

export default Header;
