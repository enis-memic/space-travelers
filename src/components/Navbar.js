import { NavLink } from 'react-router-dom';
import './Navbar.css';

const Navbar = () => (
  <nav className="navbar">
    <div className="links">
      <NavLink to="/">Rockets</NavLink>
      <NavLink to="Missions">Missions</NavLink>
      <div className="line"> </div>
      <NavLink to="Myprofile">My Profile</NavLink>
    </div>
  </nav>
);

export default Navbar;
