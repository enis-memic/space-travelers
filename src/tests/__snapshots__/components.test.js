import { MemoryRouter } from 'react-router-dom';
import {
  render,
  screen, fireEvent, within,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import store from '../../redux/store';
import Missions from '../../components/Missions';
import Navbar from '../../components/Navbar';
import Header from '../../components/Header';
import Myprofile from '../../components/Myprofile';

import '@testing-library/jest-dom/extend-expect';
import Rockets from '../../components/Rockets';

describe('Render All Components', () => {
  test('render navbar', () => {
    const { container } = render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot(`
    <div className="links">
      <a href="/">Rockets</a>
      <a href="Missions">Missions</a>
      <div className="line"> </div>
      <a href="Myprofile">My Profile</a>
    </div>`);
  });

  test('render Header', () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Header />
        </Provider>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot(`<div className="header">
    <div className="header-1">
      <img src={planet} alt="planet logo" />
      <h1 className="header-photo">Space Travelers&apos; Hub</h1>
    </div>
    <Navbar />
  </div>`);
  });

  test('render missions', () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Missions />
        </Provider>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot(`
        <table>
           <thead>
             <tr>
               <th>
                 Mission
               </th>
               <th>
                 Description
               </th>
               <th>
                 Status
               </th>
             </tr>
           </thead>
           <tbody />
         </table>
         <tr
      class="mission-list"
    >
      <td
        class="mission-title"
      />
      <td
        class="mission-description"
      />
      <td
        class="mission-status"
      >
        <button
          class="non-member"
        >
          Reserve
        </button>
      </td>
      <td
        class="mission-button"
      >
        <button
          class="Reserve"
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
    `);
  });
  test('render MyProfile', () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Myprofile />
        </Provider>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot(`
    <div className="my-profile-div">
      <div>
        <h2 className="my-profile-h2">My missions</h2>
        <table className="mprtable">
          <tbody>
            <tr key={joinedMission.mission_id}>
                 <td className="mprtd"></td>
                </tr>
          </tbody>
        </table>
      </div>
      <div className="mprdiv">
        <h2 className="my-profile-h2">My rockets</h2>
        <table className="mprtable">
          <tbody>
             <tr key={reservedRocket.id}>
                  <td className="mprtd"></td>
                </tr>
          </tbody>
        </table>
      </div>
    </div>`);
  });
});

const mockStore = configureMockStore();

describe('Rockets', () => {
  const initialState = {
    rockets: {
      rockets: [
        {
          id: '1',
          rocket_name: 'Rocket 1',
          description: 'Rocket 1 description',
          flickr_images: ['https://example.com/rocket1.jpg'],
          reserved: false,
        },
        {
          id: '2',
          rocket_name: 'Rocket 2',
          description: 'Rocket 2 description',
          flickr_images: ['https://example.com/rocket2.jpg'],
          reserved: false,
        },
        {
          id: '3',
          rocket_name: 'Rocket 3',
          description: 'Rocket 3 description',
          flickr_images: ['https://example.com/rocket2.jpg'],
          reserved: true,
        },
      ],
      fetched: true,
    },
  };

  let store;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();
  });

  test('renders Rockets component', () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(screen.getByRole('heading', { name: /Rocket 1/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Rocket 2/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Rocket 3/i })).toBeInTheDocument();
  });

  test('reserves a rocket', () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    const rocketElement = screen.getByTestId('rocket-1');
    const reserveButton = within(rocketElement).getByText(/Reserve Rocket/i);
    fireEvent.click(reserveButton);

    expect(store.getActions()).toEqual([{ type: 'rockets/bookRockets', payload: '1' }]);
  });

  test('cancels reservation', () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    const rocketElement = screen.getByTestId('rocket-3');
    const cancelButton = within(rocketElement).getByText(/Cancel Reservation/i);
    fireEvent.click(cancelButton);

    expect(store.getActions()).toEqual([{ type: 'rockets/unBookRockets', payload: '3' }]);
  });
});
