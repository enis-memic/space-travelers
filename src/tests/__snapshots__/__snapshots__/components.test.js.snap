// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render All Components render Header: <div className="header">
    <div className="header-1">
      <img src={planet} alt="planet logo" />
      <h1 className="header-photo">Space Travelers&apos; Hub</h1>
    </div>
    <Navbar />
  </div> 1`] = `
<div
  class="header"
>
  <div
    class="header-1"
  >
    <img
      alt="planet logo"
      src="planet.png"
    />
    <h1
      class="header-photo"
    >
      Space Travelers' Hub
    </h1>
  </div>
  <nav
    class="navbar"
  >
    <div
      class="links"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        Rockets
      </a>
      <a
        class=""
        href="/Missions"
      >
        Missions
      </a>
      <div
        class="line"
      >
         
      </div>
      <a
        class=""
        href="/Myprofile"
      >
        My Profile
      </a>
    </div>
  </nav>
</div>
`;

exports[`Render All Components render MyProfile: 
    <div className="my-profile-div">
      <div>
        <h2 className="my-profile-h2">My missions</h2>
        <table className="mprtable">
          <tbody>
            <tr key={joinedMission.mission_id}>
                 <td className="mprtd"></td>
                </tr>
          </tbody>
        </table>
      </div>
      <div className="mprdiv">
        <h2 className="my-profile-h2">My rockets</h2>
        <table className="mprtable">
          <tbody>
             <tr key={reservedRocket.id}>
                  <td className="mprtd"></td>
                </tr>
          </tbody>
        </table>
      </div>
    </div> 1`] = `
<div
  class="my-profile-div"
>
  <div>
    <h2
      class="my-profile-h2"
    >
      My missions
    </h2>
    <table
      class="mprtable"
    >
      <tbody />
    </table>
  </div>
  <div
    class="mprdiv"
  >
    <h2
      class="my-profile-h2"
    >
      My rockets
    </h2>
    <table
      class="mprtable"
    >
      <tbody />
    </table>
  </div>
</div>
`;

exports[`Render All Components render missions: 
        <table>
           <thead>
             <tr>
               <th>
                 Mission
               </th>
               <th>
                 Description
               </th>
               <th>
                 Status
               </th>
             </tr>
           </thead>
           <tbody />
         </table>
         <tr
      class="mission-list"
    >
      <td
        class="mission-title"
      />
      <td
        class="mission-description"
      />
      <td
        class="mission-status"
      >
        <button
          class="non-member"
        >
          Reserve
        </button>
      </td>
      <td
        class="mission-button"
      >
        <button
          class="Reserve"
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
     1`] = `
<table
  class="mtable"
>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
    </tr>
  </thead>
  <tbody
    class="tbody"
  />
</table>
`;

exports[`Render All Components render navbar: 
    <div className="links">
      <a href="/">Rockets</a>
      <a href="Missions">Missions</a>
      <div className="line"> </div>
      <a href="Myprofile">My Profile</a>
    </div> 1`] = `
<nav
  class="navbar"
>
  <div
    class="links"
  >
    <a
      aria-current="page"
      class="active"
      href="/"
    >
      Rockets
    </a>
    <a
      class=""
      href="/Missions"
    >
      Missions
    </a>
    <div
      class="line"
    >
       
    </div>
    <a
      class=""
      href="/Myprofile"
    >
      My Profile
    </a>
  </div>
</nav>
`;
